test/unittest/dds/publisher/DataWriterTests.cpp

```diff
@@ -1637,6 +1637,165 @@ TEST_F(DataWriterUnsupportedTests, UnsupportedDataWriterMethods)
+    ASSERT_NE(data_writer3, nullptr);
+
+    // Next QoS config checks that if user sets max_instances to inf and leaves max_samples by default,
+    // set_qos() should return ReturnCode_t::RETCODE_OK = 0
+    // By not using instances, this does not make any change.
+    DataWriterQos qos2 = DATAWRITER_QOS_DEFAULT;
+    DataWriter* default_data_writer1 = publisher->create_datawriter(topic, qos2);
+    ASSERT_NE(default_data_writer1, nullptr);
+
+    qos2.resource_limits().max_instances = 0;
+
+    ASSERT_EQ(ReturnCode_t::RETCODE_OK, default_data_writer1->set_qos(qos2));
+
+    // Below an ampliation of the last comprobation, for which it is proved the case of < 0 (-1),
+    // which also means infinite value.
+    // By not using instances, this does not make any change.
+    qos2.resource_limits().max_instances = -1;
+
+    ASSERT_EQ(ReturnCode_t::RETCODE_OK, default_data_writer1->set_qos(qos2));
+
+    // Next QoS config checks that if user sets max_samples < ( max_instances * max_samples_per_instance ) ,
+    // set_qos() should return ReturnCode_t::RETCODE_OK = 0
+    // By not using instances, this does not make any change.
+    qos2.resource_limits().max_samples = 4999;
+    qos2.resource_limits().max_instances = 10;
+    qos2.resource_limits().max_samples_per_instance = 500;
+
+    ASSERT_EQ(ReturnCode_t::RETCODE_OK, default_data_writer1->set_qos(qos2));
+}
+
+/*
+ * This test checks the allocation consistency when USING instances.
+ * If the topic is keyed,
+ * max_samples should be greater or equal than max_instances * max_samples_per_instance.
+ * If that condition is not met, the endpoint creation should fail.
+ * If not keyed (not using instances), the only property that is used is max_samples,
+ * thus, should not fail with the previously mentioned configuration.
+ * The following method is checked:
+ * 1. Publisher::create_datawriter
+ * 2. DataWriter::set_qos
+ */
+TEST(DataWriterTests, InstancePolicyAllocationConsistencyKeyed)
+{
+    DomainParticipant* participant =
+            DomainParticipantFactory::get_instance()->create_participant(0, PARTICIPANT_QOS_DEFAULT);
+    ASSERT_NE(participant, nullptr);
+
+    Publisher* publisher = participant->create_publisher(PUBLISHER_QOS_DEFAULT);
+    ASSERT_NE(publisher, nullptr);
+
+    TypeSupport type(new TopicDataTypeMock());
+    type.register_type(participant);
+
+    // This test pretends to use topic with instances, so the following flag is set.
+    type.get()->m_isGetKeyDefined = true;
+
+    Topic* topic = participant->create_topic("footopic", type.get_type_name(), TOPIC_QOS_DEFAULT);
+    ASSERT_NE(topic, nullptr);
+
+    // Next QoS config checks that if user sets max_instances to inf and leaves max_samples by default,
+    // create_datawriter() should not return nullptr, as the by default values are already infinite.
+    DataWriterQos qos = DATAWRITER_QOS_DEFAULT;
+    qos.resource_limits().max_instances = 0;
```

Comment:
I should remove this line because `max_instances` is 0 by default.

Also I suggest to include next cases:

- max_samples > (max_instances * max_sample_per_instances)
- max_samples = 0 and max_sample_per_instances or max_instances > 0

Reply:
27db4aa7c Refs #14958: Change in test: Test the by default qos configuration.

Reply:
Done by 4e27f9ec4 Refs 14958: DataReader tests: New test case for max_samples infinite, and ( max_instances * max_samples_per_instance ) finite.